lib/src/migrator.dart

```diff
@@ -0,0 +1,247 @@
+import 'patch.dart';
+import 'stylesheet_migration.dart';
+import 'utils.dart';
+
+/// Runs a migration of [entrypoints] and their dependencies without writing
+/// any changes to disk.
+p.PathMap<String> migrateFiles(List<String> entrypoints) =>
+    _Migrator().migrate(entrypoints);
+
+class _Migrator extends RecursiveStatementVisitor implements ExpressionVisitor {
+  /// List of all migrations for files touched by this run.
+  p.PathMap<StylesheetMigration> _migrations = p.PathMap();
+
+  /// List of migrations in progress. The last item in the current migration.
+  List<StylesheetMigration> _activeMigrations = [];
+
+  StylesheetMigration get _currentMigration =>
+      _activeMigrations.isNotEmpty ? _activeMigrations.last : null;
+
+  /// Runs the migrator on each item in [entrypoints] in turn and returns a
+  /// map of migrated contents.
+  p.PathMap<String> migrate(List<String> entrypoints) {
+    for (var entrypoint in entrypoints) {
+      _migrateStylesheet(entrypoint);
+    }
+    var results = p.PathMap<String>();
+    for (var migration in _migrations.values) {
+      results[migration.path] = migration.migratedContents;
+    }
+    return results;
+  }
+
+  /// Migrates the stylesheet at [path] if it hasn't already been migrated and
+  /// returns the StylesheetMigration instance for it regardless.
+  StylesheetMigration _migrateStylesheet(String path) {
+    path = canonicalizePath(path,
+        loadPath: _currentMigration == null
+            ? p.current
+            : p.dirname(_currentMigration.path));
+    if (_migrations.containsKey(path)) return _migrations[path];
+    var migration = StylesheetMigration(path);
+    _migrations[path] = migration;
+    _activeMigrations.add(migration);
+    visitStylesheet(migration.stylesheet);
+    _activeMigrations.remove(migration);
+    return migration;
+  }
+
+  /// Visits the children of [node] with a local scope.
+  ///
+  /// Note: The children of a stylesheet are at the root, so we should not add
+  /// a local scope.
+  @override
+  visitChildren(ParentStatement node) {
+    if (node is Stylesheet) {
+      super.visitChildren(node);
+      return;
+    }
+    _currentMigration.localScope = LocalScope(_currentMigration.localScope);
+    super.visitChildren(node);
+    _currentMigration.localScope = _currentMigration.localScope.parent;
+  }
+
+  /// Adds a namespace to any function calls that require them.
+  void visitFunctionExpression(FunctionExpression node) {
+    visitInterpolation(node.name);
+    visitArgumentInvocation(node.arguments);
+    if (node.name.asPlain == null) return;
+    var name = node.name.asPlain;
+    if (_currentMigration.localScope?.isLocalFunction(name) ?? false) return;
+    var namespace;
+    if (_currentMigration.functions.containsKey(name)) {
+      namespace =
+          _currentMigration.namespaceForNode(_currentMigration.functions[name]);
+    }
+    if (namespace == null) {
+      if (!builtInFunctionModules.containsKey(name)) return;
```

Comment:
I wonder if it might be a good idea to emit a warning here. If the migrator can't tell what a function is referring to, chances are the user invoked it wrong somehow and wants to know about it. Not something you need to do right now, but probably a good idea to add TODOs so you can easily find the places to add warnings down the line.

Reply:
`_currentMigration.functions` only contains functions defined in the global scope (whether in the current file or in one of its imports), so it's possible this could be referring to a local function.